src/validation-engine.ts

```diff
@@ -0,0 +1,228 @@
+    asyncValidationPromise.then(() => this.asyncValidationInProgressCount--);
+
+    return asyncValidationPromise;
+  }
+
+  private fireFieldAndRecordValidations = (
+    values: any
+  ): Promise<ValidationResult>[] => {
+    let fieldValidationResults: Promise<
+      ValidationResult
+    >[] = fireAllFieldsValidations(
+      values,
+      Object.keys(this.validationsPerField),
+      this.fireFieldValidations
+    );
+
+    // Now global form validations
+    if (this.validationsGlobalForm.length > 0) {
+      fieldValidationResults = [
+        ...fieldValidationResults,
+        ...this.validateGlobalFormValidations(values),
+      ];
+    }
+    return fieldValidationResults;
+  };
+
+  private buildValidationResults = (
+    validations: Promise<ValidationResult>[]
+  ): Promise<FormValidationResult> => {
+    this.asyncValidationInProgressCount++;
+    return new Promise<FormValidationResult>((resolve, reject) => {
+      // Once all the single field validations have been resolved
+      // resolve the fullFormValidatePromise
+      // TODO: Finally issue race condition unit tests
+      Promise.all(validations)
+        .then((fieldValidationResults: ValidationResult[]) => {
+          this.asyncValidationInProgressCount--;
+          const formValidationResult = buildFormValidationResult(
+            fieldValidationResults
+          );
+          resolve(formValidationResult);
+        })
+        .catch(result => {
+          this.asyncValidationInProgressCount--;
+          // Build failed validation Result
+          const errorInformation = `Uncontrolled error when validating full form, check custom validations code`;
+          console.log(errorInformation);
+          reject(result);
+        });
+    });
+  };
+
+  public fireFieldValidations = (
+    values: any,
+    key: string,
+    value: any
+  ): Promise<ValidationResult> => {
+    this.asyncValidationInProgressCount++;
+
+    const fieldValidationResultPromise = new Promise<ValidationResult>(
+      (resolve, reject) => {
+        if (!this.isFieldKeyMappingDefined(key)) {
+          this.asyncValidationInProgressCount--;
+          resolve(createDefaultValidationResult());
+        } else {
+          fireSingleFieldValidations(
+            values,
+            value,
+            this.validationsPerField[key]
+          )
+            .then((fieldValidationResult: ValidationResult) => {
+              this.asyncValidationInProgressCount--;
+              if (fieldValidationResult) {
+                fieldValidationResult.key = key;
+              }
+              resolve(fieldValidationResult);
+            })
+            .catch(result => {
+              this.asyncValidationInProgressCount--;
+              // Build failed validation Result
+              const errorInformation = `Validation Exception, field: ${key} validation fn Index: ${key}`;
+              console.log(errorInformation);
+              reject(result);
+            });
```

Comment:
The -- on the async counter is not in the finally because I had issue with jest testing and race conditions, not sure if we can find a workaround